training/components.py

```diff
@@ -167,108 +170,43 @@ def list_phase1_final_model():
-                              --learning_rate={learning_rate} \
-                              --num_warmup_steps={num_warmup_steps} \
-                              --save_samples={save_samples} \
-                              --log_level=INFO \
-                              --max_batch_len={max_batch_len} \
-                              --seed={seed} \
-                              --cpu_offload_optimizer \
-                              --cpu_offload_params_fsdp \
-                              --distributed_training_framework fsdp \
-                              --checkpoint_at_epoch
-                      command:
-                        - /bin/bash
-                        - '-c'
-                        - '--'
-                      image: {image}
-                      name: pytorch
-                      volumeMounts:
-                        - mountPath: /input_data
-                          name: input-data
-                          readOnly: true
-                        - mountPath: /input_model
-                          name: model
-                          readOnly: true
-                        - mountPath: /output
-                          name: output
-                      env:
-                        - name: NNODES
-                          value: \"{nnodes}\"
-                        - name: NPROC_PER_NODE
-                          value: \"{nproc_per_node}\"
-                        - name: XDG_CACHE_HOME
-                          value: /tmp
-                        - name: TRITON_CACHE_DIR
-                          value: /tmp
-                        - name: HF_HOME
-                          value: /tmp
-                        - name: TRANSFORMERS_CACHE
-                          value: /tmp
-                      resources:
-                        requests:
-                          "nvidia.com/gpu": {nproc_per_node}
-                        limits:
-                          "nvidia.com/gpu": {nproc_per_node}
-                  volumes:
-                    - name: input-data
-                      persistentVolumeClaim:
-                        claimName: {input_pvc_name}
-                    - name: model
-                      persistentVolumeClaim:
-                        claimName: {model_pvc_name}
-                    - name: output
-                      persistentVolumeClaim:
-                        claimName: {output_pvc_name}
-            Worker:
-              replicas: {nnodes-1}
-              restartPolicy: OnFailure
-              template:
-                metadata:
-                  annotations:
-                    sidecar.istio.io/inject: 'false'
-                spec:
-                  containers:
-                    - args:
-                        - |
-                          echo "Running phase {phase_num}"
+    resources_per_worker = {"nvidia.com/gpu": nproc_per_node}
+
+    base_image = "registry.stage.redhat.io/rhelai1/instructlab-nvidia-rhel9:1.3.1"
```

Comment:
Can we pass this as a task param? That way we should be able to do https://github.com/opendatahub-io/ilab-on-ocp/pull/231#discussion_r1880444728

```python
def pytorch_job_launcher_op(base_image=RHELAI_IMAGE,...
```

Reply:
@tumido I tried doing that but the pipelines [errors out](https://rhods-dashboard-redhat-ods-applications.apps.ocp-beta-test.nerc.mghpcc.org/experiments/ilab/5d5a1755-0df4-4a08-9a46-85b1582be2f5/runs/93655405-47e0-4c38-ba61-908b85e199c7). Basically, the constant is not available for the function defaults as well. One option could be to not set a default value for the `base_image` parameter and import the constant `RHELAI_IMAGE` in pipeline.py and pass that as a paramter while calling the `pytorch_job_launcher_op` component. WDYT?

Reply:
For the context few screenshots from @Shreyanand's run:

![image](https://github.com/user-attachments/assets/d9ebf194-9619-4ca6-b5fa-5ae33bcdf0ca)
![image](https://github.com/user-attachments/assets/d62d5c56-0a31-4de0-85ae-a7051ef8977c)
![image](https://github.com/user-attachments/assets/93e3597a-6e97-4486-a95c-0a5aa05a0b94)
![image](https://github.com/user-attachments/assets/8d5e147c-185b-4136-bd73-5f39e7c188be)

That's very interesting. The schema parses the param value properly, passes it as input yet the execution fails on unknown constant.. Looks like a KFP bug... Very weird behavior.


> One option could be to not set a default value for the base_image parameter and import the constant RHELAI_IMAGE in pipeline.py and pass that as a paramter while calling the pytorch_job_launcher_op component. WDYT?

Knowing the above I'm not convinced it would work, but please try that 🙂

Another option/workaround would be to read that value from env variable. 🙂 And this one should work. Cause KFP task runtime is not involved in parsing env variables, that goes directly from the schema.

Reply:
> Knowing the above I'm not convinced it would work, but please try that 🙂

So, it seems to be working, we just have to import the image name in the `pipeline.py` file. Weird behaviour indeed. I'll test it again and confirm with a commit.

Reply:
Added